update/Pos/Update/Poll/Logic/Base.hs

```diff
@@ -208,35 +210,39 @@ adoptBlockVersion winningBlk bv = do
 -- which currently adopted 'BlockVersion' can be applied.
 updateSlottingData
     :: (MonadError PollVerFailure m, MonadPoll m)
-    => EpochIndex -> m ()
-updateSlottingData epoch = do
-    sd@SlottingData {..} <- getSlottingData
+    => EpochIndex
+    -> m ()
+updateSlottingData epochIndex = do
     let errFmt =
             ("can't update slotting data, stored penult epoch is "%int%
              ", while given epoch is "%int%
              ")")
-    if | sdPenultEpoch + 1 == epoch -> updateSlottingDataDo sd
+
+    slottingData      <- getSlottingData
+    let nextEpochIndex = getNextEpochIndex slottingData
+    let nextEpochSD    = getNextEpochSlottingData slottingData
+
+    if | nextEpochIndex + 1 == epochIndex ->
+          updateSlottingDataDo epochIndex slottingData nextEpochSD
        -- This can happen if there was rollback of genesis block.
-       | sdPenultEpoch == epoch -> pass
+       | nextEpochIndex == epochIndex -> pass
        | otherwise ->
-           throwError $ PollInternalError $ sformat errFmt sdPenultEpoch epoch
+           throwError $ PollInternalError $ sformat errFmt nextEpochIndex epochIndex
   where
-    updateSlottingDataDo sd@SlottingData {..} = do
+    updateSlottingDataDo nextEpochIndex slottingData esd = do
         latestSlotDuration <- bvdSlotDuration <$> getAdoptedBVData
-        let epochDuration = fromIntegral epochSlots *
-                            convertUnit (esdSlotDuration sdLast)
-        let newLastStartDiff =
-                esdStartDiff sdLast + TimeDiff epochDuration
-        let newLast =
-                EpochSlottingData
-                {esdSlotDuration = latestSlotDuration, esdStartDiff = newLastStartDiff}
-        setSlottingData
-            sd
-            { sdPenultEpoch = sdPenultEpoch + 1
-            , sdPenult = sdLast
-            , sdLast = newLast
+
+        let epochDuration = fromIntegral epochSlots * convertUnit (esdSlotDuration esd)
+        let newLastStartDiff = esdStartDiff esd + TimeDiff epochDuration
+        let newLast = EpochSlottingData {
+              esdSlotDuration = latestSlotDuration
+            , esdStartDiff    = newLastStartDiff
             }
 
+        let newSlottingData = addEpochSlottingData nextEpochIndex newLast slottingData
+
+        setSlottingData newSlottingData
```

Comment:
The logic here is correct, but it's quite inefficient. In the previous PR there were these changes https://github.com/input-output-hk/cardano-sl/pull/1107/files#diff-c0fba4e2b0f4c51e5419c9671828d6b5R180
Were did they get lost? 😕